File: proteus/shockCapturing.c
#include <math.h>
#include "shockCapturing.h"

/** \file shockCapturing.c
    \ingroup shockCapturing
    @{
*/
void calculateNumericalDiffusionResGrad(int nElements_global,
                                        int nQuadraturePoints_element,
                                        int nSpace,
                                        double shockCapturingDiffusion,
                                        double* elementDiameter,
                                        double* strong_residual,
                                        double* grad_u,
                                        double* numDiff)
{
  int eN,k,I;
  double h,
    num,
    den,
    n_grad_u[nQuadraturePoints_element],
    n_R_eN,
    maxNumDiff;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      n_R_eN=0.0;
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          n_grad_u[k] = 0.0;
          for (I=0;I<nSpace;I++)
            {
              n_grad_u[k] += grad_u[eN*nQuadraturePoints_element*nSpace+k*nSpace+I]*grad_u[eN*nQuadraturePoints_element*nSpace+k*nSpace+I];
            }
          n_grad_u[k] = sqrt(n_grad_u[k]);
          num = shockCapturingDiffusion*0.5*h*fabs(strong_residual[eN*nQuadraturePoints_element+k]);
          den = n_grad_u[k];
          if (den > num*1.0e-8) 
            numDiff[eN*nQuadraturePoints_element+k] = num/den;
          else
            numDiff[eN*nQuadraturePoints_element+k]=0.0;
          n_R_eN=fmax(n_R_eN,fabs(strong_residual[eN*nQuadraturePoints_element+k]));
        }
      maxNumDiff=0.0;
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          num = shockCapturingDiffusion*0.5*h*n_R_eN;
          den = n_grad_u[k];
          if (den > num*1.0e-8)
            numDiff[eN*nQuadraturePoints_element+k] = num/den;
          else
            numDiff[eN*nQuadraturePoints_element+k]=0.0;
          maxNumDiff = fmax(maxNumDiff,numDiff[eN*nQuadraturePoints_element+k]);
        }
      for (k=0;k<nQuadraturePoints_element;k++)
         numDiff[eN*nQuadraturePoints_element+k] = maxNumDiff;
    }
}

void calculateNumericalDiffusionResGradQuad(int nElements_global,
                                            int nQuadraturePoints_element,
                                            int nSpace,
                                            double shockCapturingDiffusion,
                                            double* elementDiameter,
                                            double* strong_residual,
                                            double* grad_u,
                                            double* numDiff)
{
  int eN,k,I;
  double h,
    num,
    den,
    n_grad_u;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          n_grad_u = 0.0;
          for (I=0;I<nSpace;I++)
            {
              n_grad_u += grad_u[eN*nQuadraturePoints_element*nSpace+k*nSpace+I]*grad_u[eN*nQuadraturePoints_element*nSpace+k*nSpace+I];
            }
          num = shockCapturingDiffusion*0.5*h*fabs(strong_residual[eN*nQuadraturePoints_element+k]);
          den = sqrt(n_grad_u) + 1.0e-8;
          numDiff[eN*nQuadraturePoints_element+k] = num/den;
        }
    }
}

void calculateNumericalDiffusionEikonal(int nElements_global,
					int nQuadraturePoints_element,
					double shockCapturingDiffusion,
					double* elementDiameter,
					double* strong_residual,
					double* numDiff)
{
  int eN,k;
  double h;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          numDiff[eN*nQuadraturePoints_element+k] = shockCapturingDiffusion*0.5*h*fabs(strong_residual[eN*nQuadraturePoints_element+k]);
	  //printf("%12.5e\n",numDiff[eN*nQuadraturePoints_element+k]);
        }
    }
}

/**
   \brief Calculate the shock capturing diffusion for a Hamilton-Jacobi equation at the quadrature  points
*/
void calculateNumericalDiffusionHJ(int nElements_global,
				   int nQuadraturePoints_element,
				   char shockCapturing,
				   double shockCapturingDiffusion,
				   double* elementDiameter,
				   double* strong_residual,
				   double* mt,
				   double* H,
				   double* numDiff)
{
  int eN,k;
  double h,
    num,
    den,
    n_R_eN;
  //maxNumDiff;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      n_R_eN=0.0;
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          num = shockCapturingDiffusion*0.5*h*fabs(strong_residual[eN*nQuadraturePoints_element+k]);
          den = fabs(H[eN*nQuadraturePoints_element+k]) + 
            fabs(mt[eN*nQuadraturePoints_element+k]);
          if (den > num*1.0e-8) 
            numDiff[eN*nQuadraturePoints_element+k] = num/den;
          else
            numDiff[eN*nQuadraturePoints_element+k]=0.0;
          n_R_eN=fmax(n_R_eN,fabs(strong_residual[eN*nQuadraturePoints_element+k]));
        }
/*       maxNumDiff=0.0; */
/*       for (k=0;k<nQuadraturePoints_element;k++) */
/*         { */
/*           num = shockCapturingDiffusion*0.5*h*n_R_eN; */
/*           den = fabs(H[eN*nQuadraturePoints_element+k])+ */
/*             fabs(mt[eN*nQuadraturePoints_element+k]); */
/*           if (den > num*1.0e-8) */
/*             numDiff[eN*nQuadraturePoints_element+k] = num/den; */
/*           else */
/*             numDiff[eN*nQuadraturePoints_element+k]=0.0; */
/*           maxNumDiff = fmax(maxNumDiff,numDiff[eN*nQuadraturePoints_element+k]); */
/*         } */
/*       for (k=0;k<nQuadraturePoints_element;k++) */
/*          numDiff[eN*nQuadraturePoints_element+k] = maxNumDiff; */
    }
}

/**
   \brief Calculate the shock capturing diffusion for a Hamilton-Jacobi equation at the quadrature  points
 mwf try this with \f$H(\nabla u)\f$ in denominator
*/
void calculateNumericalDiffusionHJV2(int nElements_global,
				     int nQuadraturePoints_element,
				     char shockCapturing,
				     double shockCapturingDiffusion,
				     double* elementDiameter,
				     double* strong_residual,
				     double* mt,
				     double* H,
				     double* numDiff)
{
  int eN,k;
  double h,num,den=0.0,enorm_h;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          num = shockCapturingDiffusion*h*fabs(strong_residual[eN*nQuadraturePoints_element+
                                                           k]); 
          enorm_h = fabs(H[eN*nQuadraturePoints_element + k]);
          if(shockCapturing == '1')
            {
              den = (fabs(mt[eN*nQuadraturePoints_element+
                             k]) +
                     enorm_h)+num*1.0e-8;
            }
          else if(shockCapturing == '2')
            {
              den = sqrt(mt[eN*nQuadraturePoints_element+
                            k]
                         *
                         mt[eN*nQuadraturePoints_element+
                            k] 
                         +
                         enorm_h*enorm_h) + num*1.0e-8;
            }
          numDiff[eN*nQuadraturePoints_element+k] = num/den;
        }
    }
}
 
void calculateNumericalDiffusion_A_1(int nElements_global,
                                     int nQuadraturePoints_element,
                                     int nSpace,
                                     double shockCapturingFactor,
                                     double* elementDiameter,
                                     double* strong_residual,
                                     double* mt,
                                     double* df,
                                     double* numDiff)
{
  int eN,k,I;
  double h,num,den,enorm_df;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          num = shockCapturingFactor*h*fabs(strong_residual[eN*nQuadraturePoints_element+
                                                            k]); 
          enorm_df = 0.0;
          for(I=0;I<nSpace;I++)
            {
              enorm_df
                +=
                df[eN*nQuadraturePoints_element*nSpace+
                   k*nSpace+
                   I]
                *
                df[eN*nQuadraturePoints_element*nSpace+
                   k*nSpace+
                   I];
            }
          enorm_df = sqrt(enorm_df);
          den = (fabs(mt[eN*nQuadraturePoints_element+
                         k]) +
                 enorm_df);
          numDiff[eN*nQuadraturePoints_element+k] = num/den;
        }
    }
}
/*mwf try Juanes'  formula ...
  D_{sc,g} = C_{sc}h|R|/|u/h| */
void calculateNumericalDiffusionResGradJuanes(int nElements_global,
					      int nQuadraturePoints_element,
					      int nSpace,
					      double shockCapturingDiffusion,
					      double uSC,
					      double* elementDiameter,
					      double* strong_residual,
					      double* grad_u,
					      double* numDiff)
{
  int eN,k;
  double h,
    n_R_eN,
    num,
    den,
    maxNumDiff;
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      n_R_eN=0.0;
      for (k=0;k<nQuadraturePoints_element;k++)
	{
          num = shockCapturingDiffusion*0.5*h*fabs(strong_residual[eN*nQuadraturePoints_element+k]);
          den = fabs(uSC)/h;
          if (den > num*1.0e-8) 
            numDiff[eN*nQuadraturePoints_element+k] = num/den;
          else
            numDiff[eN*nQuadraturePoints_element+k]=0.0;
          n_R_eN=fmax(n_R_eN,fabs(strong_residual[eN*nQuadraturePoints_element+k]));
        }
      maxNumDiff=0.0;
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          num = shockCapturingDiffusion*0.5*h*n_R_eN;
          den = fabs(uSC)/h;
          if (den > num*1.0e-8)
            numDiff[eN*nQuadraturePoints_element+k] = num/den;
          else
            numDiff[eN*nQuadraturePoints_element+k]=0.0;
          maxNumDiff = fmax(maxNumDiff,numDiff[eN*nQuadraturePoints_element+k]);
        }
      for (k=0;k<nQuadraturePoints_element;k++)
         numDiff[eN*nQuadraturePoints_element+k] = maxNumDiff;
    }
}

/****
     supposed to be nu_i = nu_c h^{2-beta}|R_i|
 ***/

void calculateNumericalDiffusionJaffre(int nElements_global,
				       int nQuadraturePoints_element,
				       int nSpace,
				       double shockCapturingDiffusion,
				       double beta,
				       double* elementDiameter,
				       double* strong_residual,
				       double* grad_u,
				       double* numDiff)
{
  int eN,k,I;
  double h,
    num,
    hfactor,
    n_grad_u[nQuadraturePoints_element],
    n_R_eN,
    maxNumDiff;
  
  for(eN=0;eN<nElements_global;eN++)
    {
      h = elementDiameter[eN];
      hfactor = pow(h,2.0-beta);
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          num = shockCapturingDiffusion*hfactor*fabs(strong_residual[eN*nQuadraturePoints_element+k]);
	  numDiff[eN*nQuadraturePoints_element+k] = num;
        }
      
      maxNumDiff=0.0;
      for (k=0;k<nQuadraturePoints_element;k++)
        {
          maxNumDiff = fmax(maxNumDiff,numDiff[eN*nQuadraturePoints_element+k]);
        }
      for (k=0;k<nQuadraturePoints_element;k++)
         numDiff[eN*nQuadraturePoints_element+k] = maxNumDiff;
    }
}


/** @} */

File: proteus/ShockCapturing.py
"""
A class hierarchy for shock capturing diffusion methods

.. inheritance-diagram:: proteus.ShockCapturing
   :parts: 1
"""
import numpy
from . import cshockCapturing
class ShockCapturing_base(object):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True):
        self.nc = coefficients.nc
        self.nd = nd
        self.components=list(range(self.nc))
        self.shockCapturingFactor=shockCapturingFactor
        self.lag=lag
        #mwf had to put in da_sge, df_sge
        self.coefficients = coefficients
    def initializeElementQuadrature(self,mesh,t,cq):
        self.mesh=mesh
        self.numDiff=[]
        self.numDiff_last=[]
        for ci in range(self.nc):
            if self.lag:
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(numpy.zeros(cq[('u',ci)].shape,'d'))
            else:
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(cq[('numDiff',ci,ci)])
        #mwf what if just want shock capturing and no stabilization?
        #need to put in lag steps for da_sge and df_sge below as well if not
        #done by Subgrid error
        for ci in range(self.nc):
            for cj in range(self.nc):
                if ('df',ci,cj) in cq and ('df_sge',ci,cj) not in cq:
                    cq[('df_sge',ci,cj)]=cq[('df',ci,cj)]
        for ci,ckDict in self.coefficients.diffusion.items():
            for ck,cjDict in ckDict.items():
                if ('grad(phi)_sge',ck) not in cq:
                    cq[('grad(phi)_sge',ck)]=cq[('grad(phi)',ck)]
                for cj in list(cjDict.keys()):
                    if ('dphi_sge',ck,cj) not in cq:
                        cq[('dphi_sge',ck,cj)]=cq[('dphi',ck,cj)]
                    if ('da_sge',ci,ck,cj) not in cq:
                        cq[('da_sge',ci,ck,cj)]=cq[('da',ci,ck,cj)]

    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            self.numDiff[ci].flat[:] = self.shockCapturingFactor
    def updateShockCapturingHistory(self):
        if self.lag:
            for ci in range(self.nc):
                self.numDiff_last[ci][:] = self.numDiff[ci]

ConstantDiffusion_SC = ShockCapturing_base

class ResGrad_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True):
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusionResGrad(self.shockCapturingFactor,
                                                               self.mesh.elementDiametersArray,
                                                               q[('pdeResidual',ci)],
                                                               q[('grad(u)',ci)],
                                                               self.numDiff[ci])
class ResGradFFDarcy_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True):
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
    def calculateNumericalDiffusion(self,q):
        cshockCapturing.calculateNumericalDiffusionResGrad(self.shockCapturingFactor,
                                                           self.mesh.elementDiametersArray,
                                                           q[('pdeResidual',0)],
                                                           q[('grad(u)',0)],
                                                           self.numDiff[0])

class ResGradQuad_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True,gradLag=True):
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.debug=False
        self.gradLag = gradLag
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            if self.debug:
                if numpy.isnan(q[('pdeResidual',ci)]).any(): #
                    import pdb
                    print("NaN's in res")
                    pdb.set_trace()
            cshockCapturing.calculateNumericalDiffusionResGradQuad(self.shockCapturingFactor,
                                                                   self.mesh.elementDiametersArray,
                                                                   q[('pdeResidual',ci)],
                                                                   q[('grad(u)',ci)],
                                                                   self.numDiff[ci])
            if self.debug:
                if numpy.isnan(self.numDiff[ci]).any():
                    import pdb
                    print("NaN's in numDiff")
                    pdb.set_trace()

class Eikonal_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True):
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.debug=False
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusionEikonal(self.shockCapturingFactor,
                                                               self.mesh.elementDiametersArray,
                                                               q[('pdeResidual',ci)],
                                                               self.numDiff[ci])

class ScalarAdvection_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.5,lag=True):
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusion_A_1(self.shockCapturingFactor,
                                                            self.mesh.elementDiametersArray,
                                                            q[('pdeResidual',ci)],
                                                            q[('mt',ci)],
                                                            q[('df',ci,ci)],
                                                            self.numDiff[ci])

class HamiltonJacobi_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFlag='1',shockCapturingFactor=0.5,lag=True):
        self.shockCapturingFlag=shockCapturingFlag=shockCapturingFlag
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusionHJ(self.shockCapturingFlag,
                                                          self.shockCapturingFactor,
                                                          self.mesh.elementDiametersArray,
                                                          q[('pdeResidual',ci)],
                                                          q[('mt',ci)],
                                                          q[('H',ci)],
                                                          self.numDiff[ci])
class HamiltonJacobiJaffre_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFlag='1',shockCapturingFactor=0.5,lag=True,betaPower=0.1):
        self.shockCapturingFlag=shockCapturingFlag=shockCapturingFlag
        self.beta=betaPower
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusionJaffre(self.shockCapturingFactor,
                                                              self.beta,
                                                              self.mesh.elementDiametersArray,
                                                              q[('pdeResidual',ci)],
                                                              q[('dH',ci,ci)],
                                                              self.numDiff[ci])
class JaffreGradU_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFlag='1',shockCapturingFactor=0.5,lag=True,betaPower=0.1):
        self.shockCapturingFlag=shockCapturingFlag=shockCapturingFlag
        self.beta=betaPower
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusionJaffre(self.shockCapturingFactor,
                                                              self.beta,
                                                              self.mesh.elementDiametersArray,
                                                              q[('pdeResidual',ci)],
                                                              q[('grad(u)',ci)],
                                                              self.numDiff[ci])

class ResGradJuanes_SC(ShockCapturing_base):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,uSC=1.0,lag=True):
        ShockCapturing_base.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.uSC = uSC
    def calculateNumericalDiffusion(self,q):
        for ci in range(self.nc):
            cshockCapturing.calculateNumericalDiffusionResGradJuanes(self.shockCapturingFactor,
                                                                     self.uSC,
                                                                     self.mesh.elementDiametersArray,
                                                                     q[('pdeResidual',ci)],
                                                                     q[('grad(u)',ci)],
                                                                     self.numDiff[ci])

#examples of shock capturing with lagging allowed after certain number of steps
class ResGradDelayLag_SC(ResGrad_SC):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True,nStepsToDelay=None):
        ResGrad_SC.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.nStepsToDelay = nStepsToDelay
        self.nSteps=0
    def initializeElementQuadrature(self,mesh,t,cq):
        self.mesh=mesh
        self.numDiff=[]
        self.numDiff_last=[]
        self.cq_numDiff=[]
        for ci in range(self.nc):
            if self.lag:
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(numpy.zeros(cq[('u',ci)].shape,'d'))
            elif self.lag == False and self.nStepsToDelay is not None:
                self.cq_numDiff.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(cq[('numDiff',ci,ci)])
            else:
                self.numDiff.append(cq[('numDiff',ci,ci)])
        #mwf what if just want shock capturing and no stabilization?
        #need to put in lag steps for da_sge and df_sge below as well if not
        #done by Subgrid error
        for ci in range(self.nc):
            for cj in range(self.nc):
                if ('df',ci,cj) in cq and ('df_sge',ci,cj) not in cq:
                    cq[('df_sge',ci,cj)]=cq[('df',ci,cj)]
        for ci,ckDict in self.coefficients.diffusion.items():
            for ck,cjDict in ckDict.items():
                if ('grad(phi)_sge',ck) not in cq:
                    cq[('grad(phi)_sge',ck)]=cq[('grad(phi)',ck)]
                for cj in list(cjDict.keys()):
                    if ('dphi_sge',ck,cj) not in cq:
                        cq[('dphi_sge',ck,cj)]=cq[('dphi',ck,cj)]
                    if ('da_sge',ci,ck,cj) not in cq:
                        cq[('da_sge',ci,ck,cj)]=cq[('da',ci,ck,cj)]
    #
    def updateShockCapturingHistory(self):
        self.nSteps += 1
        if self.lag:
            for ci in range(self.nc):
                self.numDiff_last[ci][:] = self.numDiff[ci]
        if self.lag == False and self.nStepsToDelay is not None and self.nSteps > self.nStepsToDelay:
            self.lag = True
            self.numDiff = []
            self.numDiff_last=[]
            for ci in range(self.nc):
                self.numDiff_last.append(self.cq_numDiff[ci])
                self.numDiff.append(numpy.zeros(self.cq_numDiff[ci].shape,'d'))

class ResGradQuadDelayLag_SC(ResGradQuad_SC):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True,nStepsToDelay=None,isotropic=False):
        ResGradQuad_SC.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.nStepsToDelay = nStepsToDelay
        self.nSteps=0
        self.isotropic=isotropic
    def initializeElementQuadrature(self,mesh,t,cq):
        self.mesh=mesh
        self.numDiff=[]
        self.numDiff_last=[]
        for ci in range(self.nc):
            if self.lag:
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(numpy.zeros(cq[('u',ci)].shape,'d'))
            elif self.lag == False and self.nStepsToDelay is not None:
                self.numDiff.append(cq[('numDiff',ci,ci)])
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
            else:
                self.numDiff.append(cq[('numDiff',ci,ci)])
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
    def updateShockCapturingHistory(self):
        self.nSteps += 1

        if self.isotropic:
            for ci in range(1,self.nc):
                self.numDiff[0][:] = numpy.maximum(self.numDiff[0],self.numDiff[ci])
            for ci  in range(1,self.nc):
                self.numDiff[ci][:] = self.numDiff[0]
        if self.lag:
            for ci in range(self.nc):
                self.numDiff_last[ci][:] = self.numDiff[ci]
        if self.lag == False and self.nStepsToDelay is not None and self.nSteps > self.nStepsToDelay:
            self.lag = True
            self.numDiff=[]
            for ci in range(self.nc):
                self.numDiff.append(self.numDiff_last[ci].copy())

class NavierStokes_SC(ResGradQuad_SC):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True,nStepsToDelay=None):
        ResGradQuad_SC.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.nStepsToDelay = nStepsToDelay
        self.nSteps=0
    def calculateNumericalDiffusion(self,q):
        for ci in range(1,self.nc):
            if numpy.isnan(q[('pdeResidual',ci)]).any():
                import pdb
                print("NaN's in res")
                pdb.set_trace()
            cshockCapturing.calculateNumericalDiffusionResGradQuad(self.shockCapturingFactor,
                                                                   self.mesh.elementDiametersArray,
                                                                   q[('pdeResidual',ci)],
                                                                   q[('grad(u)',ci)],
                                                                   self.numDiff[ci])
            if numpy.isnan(self.numDiff[ci]).any():
                import pdb
                print("NaN's in numDiff")
                pdb.set_trace()
    def initializeElementQuadrature(self,mesh,t,cq):
        self.mesh=mesh
        self.numDiff={}
        self.numDiff_last={}
        self.cq_numDiff={}
        for ci in range(1,self.nc):
            if self.lag:
                self.numDiff_last[ci]=cq[('numDiff',ci,ci)]
                self.numDiff[ci]=numpy.zeros(cq[('u',ci)].shape,'d')
            elif self.lag == False and self.nStepsToDelay is not None:
                self.cq_numDiff[ci]=cq[('numDiff',ci,ci)]
                self.numDiff[ci]=cq[('numDiff',ci,ci)]
            else:
                self.numDiff[ci]=cq[('numDiff',ci,ci)]
    def updateShockCapturingHistory(self):
        self.nSteps += 1
        if self.lag:
            for ci in range(1,self.nc):
                self.numDiff_last[ci][:] = self.numDiff[ci]
        if self.lag == False and self.nStepsToDelay is not None and self.nSteps > self.nStepsToDelay:
            self.lag = True
            self.numDiff = []
            self.numDiff_last=[]
            for ci in range(1,self.nc):
                self.numDiff_last[ci]=self.cq_numDiff[ci]
                self.numDiff[ci]=numpy.zeros(self.cq_numDiff[ci].shape,'d')

class NavierStokes_SC_opt(ResGradQuad_SC):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True,nStepsToDelay=None):
        ResGradQuad_SC.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.nStepsToDelay = nStepsToDelay
        self.nSteps=0
    def calculateNumericalDiffusion(self,q):
        for ci in range(1,self.nc):
            if numpy.isnan(q[('pdeResidual',ci)]).any():
                import pdb
                print("NaN's in res")
                pdb.set_trace()
            cshockCapturing.calculateNumericalDiffusionResGradQuad(self.shockCapturingFactor,
                                                                   self.mesh.elementDiametersArray,
                                                                   q[('pdeResidual',ci)],
                                                                   q[('grad(u)',ci)],
                                                                   self.numDiff[ci])
            if numpy.isnan(self.numDiff[ci]).any():
                import pdb
                print("NaN's in numDiff")
                pdb.set_trace()
    def initializeElementQuadrature(self,mesh,t,cq):
        self.mesh=mesh
        self.numDiff={}
        self.numDiff_last={}
        for ci in range(1,self.nc):
            if self.lag:
                #note: _opt  does  the  lagging backwards from above
                self.numDiff_last[ci]=numpy.zeros(cq[('u',ci)].shape,'d')
                self.numDiff[ci]=cq[('numDiff',ci,ci)]
            else:
                self.numDiff_last[ci]=cq[('numDiff',ci,ci)]
                self.numDiff[ci]=cq[('numDiff',ci,ci)]
    def updateShockCapturingHistory(self):
        for ci in range(1,self.nc):
            if self.lag:
                self.numDiff_last[ci][:]=self.numDiff[ci]

class ResGradFFDarcyDelayLag_SC(ResGradFFDarcy_SC):
    def __init__(self,coefficients,nd,shockCapturingFactor=0.25,lag=True,nStepsToDelay=None):
        ResGradFFDarcy_SC.__init__(self,coefficients,nd,shockCapturingFactor,lag)
        self.nStepsToDelay = nStepsToDelay
        self.nSteps=0
    def initializeElementQuadrature(self,mesh,t,cq):
        self.mesh=mesh
        self.numDiff=[]
        self.numDiff_last=[]
        self.cq_numDiff=[]
        for ci in range(self.nc):
            if self.lag:
                self.numDiff_last.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(numpy.zeros(cq[('u',ci)].shape,'d'))
            elif self.lag == False and self.nStepsToDelay is not None:
                self.cq_numDiff.append(cq[('numDiff',ci,ci)])
                self.numDiff.append(cq[('numDiff',ci,ci)])
            else:
                self.numDiff.append(cq[('numDiff',ci,ci)])
    def updateShockCapturingHistory(self):
        self.nSteps += 1
        if self.lag:
            for ci in range(self.nc):
                self.numDiff_last[ci][:] = self.numDiff[ci]
        if self.lag == False and self.nStepsToDelay is not None and self.nSteps > self.nStepsToDelay:
            self.lag = True
            self.numDiff = []
            self.numDiff_last=[]
            for ci in range(self.nc):
                self.numDiff_last.append(self.cq_numDiff[ci])
                self.numDiff.append(numpy.zeros(self.cq_numDiff[ci].shape,'d'))
